sql/mysqld.cc

```diff
@@ -5587,6 +5591,70 @@ static void test_lc_time_sz()
 }
 #endif//DBUG_OFF
 
+static void check_upgrade(void)
+{
+  File info_file;
+  char upgrade_from_version[1024]= {0};
+  size_t length;
+  ulong version_id;
+
+  uint major, minor, version;
+  char *point= upgrade_from_version, *end_point, *pos;
+
+  DBUG_ENTER("check_upgrade");
+
+  if (upgrade_info_file != 0)
+  {
+    fn_format(upgrade_info_file, "mysql_upgrade_info", mysql_real_data_home_ptr, "", MYF(0));
+    DBUG_PRINT("info", ("mysql_upgrade_info path: %s", mysql_real_data_home_ptr));
+  }
+
+  if ((info_file= my_open(upgrade_info_file, O_RDONLY, MYF(0))) < 0)
+  {
+    sql_print_information("Cannot open mysql_upgrade_info file: %s. Please run mariadb-upgrade.", upgrade_info_file);
+    DBUG_VOID_RETURN;
+  }
+
+  if (my_lock(info_file, F_RDLCK, 0, 1, MYF(0)))
+    goto end; // If Another process is holding a write lock avoid checking
+
+  (void) my_seek(info_file, 0, SEEK_SET, MYF(0));
+  /* We have -3 here to make conversion below safe (similar to mysql_upgrade) */
+  length= my_read(info_file, (uchar*) upgrade_from_version,
+                  sizeof(upgrade_from_version)-3, MYF(0));
+
+  /* Remove possible \ŋ that may end in output */
+  if ((pos= strchr(upgrade_from_version, '\n')))
+    *pos= 0;
+
+  if (!length)
+  {
+    sql_print_warning("mysql_upgrade_info file: %s is empty. Please run mariadb-upgrade.", upgrade_info_file);
+    goto end;
+  }
+
+  major=   (uint) strtoul(point, &end_point, 10);   point=end_point+1;
+  minor=   (uint) strtoul(point, &end_point, 10);   point=end_point+1;
+  version= (uint) strtoul(point, &end_point, 10);
+
+  if (major == minor && minor == version && version == 0)
```

Comment:
so this is a condition if they are all 0. Not the most obvious of expression. `!(major || minor || version)`  maybe.